src/Microsoft.Framework.PackageManager/Utils/NuGetPackageUtils.cs

```diff
@@ -27,18 +27,33 @@ internal static class NuGetPackageUtils
             var targetPath = packagePathResolver.GetInstallPath(library.Name, library.Version);
             var targetNuspec = packagePathResolver.GetManifestFilePath(library.Name, library.Version);
             var targetNupkg = packagePathResolver.GetPackageFilePath(library.Name, library.Version);
-            var hashPath = packagePathResolver.GetHashPath(library.Name, library.Version);
+            var targetHashPath = packagePathResolver.GetHashPath(library.Name, library.Version);
 
             // Acquire the lock on a nukpg before we extract it to prevent the race condition when multiple
             // processes are extracting to the same destination simultaneously
-            await ConcurrencyUtilities.ExecuteWithFileLocked(targetNupkg, async createdNewLock =>
+            await ConcurrencyUtilities.ExecuteWithFileLocked(targetNupkg, async _ =>
             {
-                // If this is the first process trying to install the target nupkg, go ahead
-                // After this process successfully installs the package, all other processes
-                // waiting on this lock don't need to install it again.
-                if (createdNewLock && !File.Exists(targetNupkg))
+                string packageHash;
+                using (var sha512 = SHA512.Create())
                 {
-                    information.WriteLine($"Installing {library.Name.Bold()} {library.Version}");
+                    packageHash = Convert.ToBase64String(sha512.ComputeHash(stream));
+                }
+
+                var actionName = "Installing";
+                var installedPackageHash = string.Empty;
+                if (File.Exists(targetHashPath))
+                {
+                    installedPackageHash = File.ReadAllText(targetHashPath);
+                    actionName = "Overwriting";
+                }
+
+                if (string.Equals(packageHash, installedPackageHash, StringComparison.Ordinal))
+                {
+                    information.WriteLine($"{library.Name.Bold()}.{library.Version} already exists");
```

Comment:
:+1: 

` ...already exists and won't be overwritten because it is identical and we don't overwrite identical files except for the SHA files` :smiley: